felix/rules/policy_test.go

```diff
@@ -18,15 +18,23 @@ import (
 	"github.com/projectcalico/calico/felix/environment"
 	. "github.com/projectcalico/calico/felix/rules"
 
-	. "github.com/onsi/ginkgo"
-	. "github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/ginkgo/v2"
 	. "github.com/onsi/gomega"
 
 	"github.com/projectcalico/calico/felix/ipsets"
 	"github.com/projectcalico/calico/felix/iptables"
 	"github.com/projectcalico/calico/felix/proto"
 )
 
+func describeTable(s string, f interface{}, entries ...TableEntry) {
```

Comment:
Hm, what is this for?

Reply:
This is because golang is dumb in the sense that if you have `func(blah ...interface{})`  and you give it `s string, f interface{}, entries ...TableEntry` (which `DescribeTable` used to take) it does not figure out that those are all `interfaces{}` so we have to make it a single slice of `interface{}` and this level of indirection was the least disrupting. But happy to use a simpler way if there is :man_shrugging:

Reply:
Ah, I think I see.

The v2 example shows a format that looks really similar: https://pkg.go.dev/github.com/onsi/ginkgo/v2#DescribeTable

But in our case, we want to reuse tableEntries in multiple tables and golang doesn't like that `tableEntries...` and `f` aren't the same type?

My main worry is that we're going to end up having this pattern copied throughout our test code and it's a little sad, especially if it ends up being done in places it's not necessary (e.g., via copy / pasting your example). 

I wonder if we were starting from scratch, if we'd do something like this instead:


```go

// Logic for test one.
testF1 := func()  {}

// Logic for test two.
testF2 := func()  {}

// Common test data for all table driven tests.
testData := []interface{ . . .}

DescribeTable("Desc T1", append([]interface{testF1}, testData...))
DescribeTable("Desc T2", append([]interface{testF2}, testData...))

```

Obviously that's a bit more work than this wrapper is, but might be a nicer long term approach. Still kinda sad.

Either way I'm not too bothered - the choice is yours - but I hope that this can be constrained to this one place rather than becoming too widespread a pattern... If we do need to make this a more common pattern, then I think we should do it properly with a public function so that we can do it consistently throughout all of our table driven tests rather than needing to redefine this in multiple test files.

As an aside, this test file seems odd anyway since it has a `DescribeTable` within a `Describe`, which is error prone and is actually called out explicitly as something to avoid in the Ginkgo docs, it seems, so this test file could use a small refactor anyway :grin:

Reply:
Yeah, I actually do this in at least on of the libcalico-go files. The others are only a matter of getting rid of the array. Thought about this here too, I will have a second look. It is so far just this file.